### ISSUE TITLE
discuss: Release arrow-udf 0.4.0

### ISSUE THREAD

**Issue by Xuanwo** [arrow-udf/arrow-udf#65]
Hello, everyone. I'm starting this thread to discuss the release of arrow-udf 0.2.1. This release pins the arrow version to `52` to ensure our users can build upon it.

- https://github.com/arrow-udf/arrow-udf/pull/63

As discussed in https://github.com/arrow-udf/arrow-udf/pull/63, we will bump to arrow 53 to catch up the arrow train. The next release will be `0.3.0`, what do you think?

cc @wangrunji0408 @sundy-li @stdrc @xxchan

**Comment by xxchan** [comment]
I'm thinking: what about using feature flags, instead of a range to specify the arrow version?

**Comment by Xuanwo** [comment]
> I'm thinking: what about using feature flags, instead of a range to specify the arrow version?

Since arrow-rs is planning a stable release every three months, I believe we can simply pin the arrow version and update it every three months as well.

**Comment by xxchan** [comment]
Only supporting latest version SGTM 😄

**Comment by Xuanwo** [comment]
Great, I'll work on this release.

**Comment by xxchan** [comment]
BTW, should we consider re-export the arrow types? So that users don't need to add arrow in their `Cargo.toml`.

**Comment by Xuanwo** [comment]
> BTW, should we consider re-export the arrow types? So that users don't need to add arrow in their `Cargo.toml`. This can make upgrading `arrow-udf` version less painful.

Personally, I don't like this idea since users rarely use `arrow-udf` alone. Most of the time, they already have some data from or into Arrow.

**Comment by xxchan** [comment]
IIUC, whenever a crate uses a dependency's type in the public API, it would be better to re-export the dependency. Of course the largest benefit is ease of handling multiple versions.

Currently we have things like:
```
arrow-array-iceberg = { package = "arrow-array", version = "52" }
arrow-schema-iceberg = { package = "arrow-schema", version = "52" }
arrow-buffer-iceberg = { package = "arrow-buffer", version = "52" }
arrow-cast-iceberg = { package = "arrow-cast", version = "52" }
arrow-array = "50"
arrow-arith = "50"
arrow-cast = "50"
arrow-schema = "50"
arrow-buffer = "50"
arrow-flight = "50"
arrow-select = "50"
arrow-ord = "50"
arrow-row = "50"
arrow-array-deltalake = { package = "arrow-array", version = "48.0.1" }
arrow-buffer-deltalake = { package = "arrow-buffer", version = "48.0.1" }
arrow-cast-deltalake = { package = "arrow-cast", version = "48.0.1" }
arrow-schema-deltalake = { package = "arrow-schema", version = "48.0.1" }
```
This is quite awkward.

Some posts about this:
- https://www.lurklurk.org/effective-rust/re-export.html
- https://brokenco.de/2023/07/26/rust-re-export.html

**Comment by Xuanwo** [comment]
Could you elaborate more on how "re-export the dependency" handling the multi-version issues? For example, I have a `arrow-array = "50"`'s `Array`, how to feed it in to iceberg's arroy?

**Comment by xxchan** [comment]
It doesn't help in this case (i.e., when you have a specific version's type, and want to pass it to a lib using the dep). 

The case it improves is: when you have code to interact with `arrow-udf` or `iceberg`, you construct it with `arrow-udf::arrow::Array`, or `iceberg::arrow::Array`. There's no inter-op between `arrow-udf::arrow::Array`, and `iceberg::arrow::Array` (or your own `arrow::Array` used in your code).

In our case, it's exactly like this:
```
RisingWave Array   --- IcebergArrowConvert --> iceberg Array
				 
				  |--- UdfArrowConvert     --> udf Array
```
If each crate re-exports `arrow`, we don't need to improve arrow and worry about the versions.

Example in this post may be simpler to understand: https://www.lurklurk.org/effective-rust/re-export.html
It uses `dep_lib::rand` to construct `dep_lib::rand::Rng` (rand V0.7), instead of `rand::Rng` (rand V0.8).

---

Basically, the above discussion assumes that multiple versions are to some extent unavoidable.

If you want to ensure only one version is used across your project, that's orthogonal to the problem. And actually re-export doesn't hurt in any case.

**Comment by xxchan** [comment]
TLDR is that: the whole point is things like `arrow-array-iceberg` doesn't make sense to me, and can be solved by re-export.

**Comment by Xuanwo** [comment]
> TLDR is that: the whole point is things like `arrow-array-iceberg` doesn't make sense to me, and can be solved by re-export.

Hi, I apologize for closing this issue by merging https://github.com/arrow-udf/arrow-udf/pull/68. Would you like to create a new issue if you believe it's worthwhile to expose arrow types?
